include/gatekeeper_lls.h

```diff
@@ -19,6 +19,183 @@
 #ifndef _GATEKEEPER_LLS_H_
 #define _GATEKEEPER_LLS_H_
 
-int run_lls(void);
+#include <rte_timer.h>
+
+#include "gatekeeper_mailbox.h"
+#include "gatekeeper_net.h"
+
+/*
+ * Maximum key length (in bytes) for an LLS map. It should be set
+ * to the maximum key length needed. Currently, this is set to 16 for
+ * the number of bytes in an IPv6 address.
+ */
+#define LLS_MAX_KEY_LEN (16)
+
+/* Number of records that a LLS cache can hold. */
+#define LLS_CACHE_RECORDS (1024)
+
+/* Requests that can be made to a cache. */
+enum lls_req_ty {
+	/* Do a one-time resolution; do not receive further updates. */
+	LLS_REQ_RESOLVE,
+	/* Request to receive a resolution and any updates for this record. */
+	LLS_REQ_SUBSCRIBE,
+	/* Remove a previously-submitted hold. */
+	LLS_REQ_PUT,
+	/* Modify a record, possibly with a new address or port. */
+	LLS_REQ_MODIFY,
+};
+
+/* Replies that come from a cache. */
+enum lls_reply_ty {
+	/* The reply represents the first time the block is being notified. */
+	LLS_REPLY_RESOLUTION,
+	/* The reply represents an update to a previous resolution. */
+	LLS_REPLY_UPDATE,
+	/*
+	 * The reply is a notification that the hold is
+	 * removed, so the requester can free state as needed.
+	 */
+	LLS_REPLY_FREE,
+};
+
+/* Map that is returned to blocks that request resolution. */
+struct lls_map {
+	/* Ethernet address of this map. */
+	struct ether_addr ha;
+
+	/* Port on which this map exists. */
+	uint8_t           port_id;
+
+	/* Whether this map has been marked as stale. */
+	int               stale;
+
+	/* IP address for this map, in network ordering. */
+	uint8_t           ip_be[LLS_MAX_KEY_LEN];
+};
+
+/*
+ * Format of callback function for requesting LLS maps.
+ *
+ * The LLS block invokes the callbacks, so each block should ensure
+ * that the callback function deals with any race conditions and
+ * is aware that the blocks may reside in different NUMA nodes.
+ */
+typedef void (*lls_req_cb)(struct lls_map const *map, void *arg,
```

Comment:
The intention here is to have the compiler enforce that the data of the map should not be changed, since this pointer is passed to potentially many callbacks. These callbacks should not be allowed the change the map itself.